common/web/types/src/util/util.ts

```diff
@@ -101,3 +101,153 @@ toOneChar(value: string) : number {
   }
   return value.codePointAt(0);
 }
+
+export function describeCodepoint(ch : number) : string {
+  let s;
+  const p = getProblem(ch);
+  if (p != null) {
+    // for example: 'PUA (U+E010)'
+    s = p;
+  } else {
+    // for example: '"a" (U+61)'
+    s = `"${String.fromCodePoint(ch)}"`;
+  }
+  return `${s} (U+${Number(ch).toString(16).toUpperCase()})`;
+}
+
+export enum BadStringType {
+  pua = 'PUA',
+  unassigned = 'Unassigned',
+  illegal = 'Illegal',
+};
+
+// Following from kmx_xstring.h / .cpp
+
+const Uni_LEAD_SURROGATE_START = 0xD800;
+const Uni_LEAD_SURROGATE_END = 0xDBFF;
+const Uni_TRAIL_SURROGATE_START = 0xDC00;
+const Uni_TRAIL_SURROGATE_END = 0xDFFF;
+const Uni_SURROGATE_START = Uni_LEAD_SURROGATE_START;
+const Uni_SURROGATE_END = Uni_TRAIL_SURROGATE_END;
+const Uni_FD_NONCHARACTER_START = 0xFDD0;
+const Uni_FD_NONCHARACTER_END = 0xFDEF;
+const Uni_FFFE_NONCHARACTER = 0xFFFE;
+const Uni_PLANE_MASK = 0x1F0000;
+const Uni_MAX_CODEPOINT = 0x10FFFF;
+
+/**
+ * @brief True if a lead surrogate
+ * \def Uni_IsSurrogate1
+ */
+function Uni_IsSurrogate1(ch : number) {
+  return ((ch) >= Uni_LEAD_SURROGATE_START && (ch) <= Uni_LEAD_SURROGATE_END);
+}
+/**
+ * @brief True if a trail surrogate
+ * \def Uni_IsSurrogate2
+ */
+function Uni_IsSurrogate2(ch : number) {
+  return ((ch) >= Uni_TRAIL_SURROGATE_START && (ch) <= Uni_TRAIL_SURROGATE_END);
+}
+
+/**
+ * @brief True if any surrogate
+ * \def UniIsSurrogate
+*/
+function Uni_IsSurrogate(ch : number) {
+  return (Uni_IsSurrogate1(ch) || Uni_IsSurrogate2(ch));
+}
+
+function Uni_IsEndOfPlaneNonCharacter(ch : number) {
+  return (((ch) & Uni_FFFE_NONCHARACTER) == Uni_FFFE_NONCHARACTER); // matches FFFF or FFFE
+}
+
+function Uni_IsNoncharacter(ch : number) {
+  return (((ch) >= Uni_FD_NONCHARACTER_START && (ch) <= Uni_FD_NONCHARACTER_END) || Uni_IsEndOfPlaneNonCharacter(ch));
+}
+
+function Uni_InCodespace(ch : number) {
+  return ((ch) <= Uni_MAX_CODEPOINT);
+};
```

Comment:
```suggestion
function Uni_InCodespace(ch : number) {
  return (ch >= 0 && ch <= Uni_MAX_CODEPOINT);
};
```

Reply:
right, because it was ported to an environment that doesn't have unsigned…